fix: parsing of .npmrc file

emmenko opened the pull request:
Fixes #114 

I decided to use the `npm/ini` package to take care of the parse/stringify part of the `.npmrc` config (which is what NPM itself uses). This avoids us having to deal with whitespaces etc.

To make it easier to test, I extracted the logic into a function in a separate file `npmUtils`.

emmenko committed:
fix: parsing of .npmrc file

emmenko commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,50 @@
+    console.log(`Found existing user .npmrc file at ${userNpmrcPath}`);
+
+    // Parse the `.npmrc` content using the `npm/ini` package.
+    const npmConfig = ini.parse(fs.readFileSync(userNpmrcPath, "utf-8"));
```
We use `ini.parse` to take care of parsing the config (all whitespaces are stripped out)

emmenko commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,50 @@
+    // Parse the `.npmrc` content using the `npm/ini` package.
+    const npmConfig = ini.parse(fs.readFileSync(userNpmrcPath, "utf-8"));
+
+    let hasAuthToken = false;
+    for (const [key, value] of Object.entries(npmConfig)) {
+      if (/\/\/(.*)authToken$/.test(key) && Boolean(value)) {
+        console.log("The .npmrc file has an authToken");
+        hasAuthToken = true;
+      }
+    }
```
This is the "new logic" to test the presence of the `authToken`. The regex also can be simpler.

emmenko commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,50 @@
+      console.log(
+        "The .npmrc file does not have an authToken defined, creating one using the `NPM_TOKEN` environment variable"
+      );
+      if (!processEnv.NPM_TOKEN) {
```
I thought it would make sense to throw an error if the `NPM_TOKEN` is not present.

emmenko commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,50 @@
+          "Missing NPM authToken. Please make sure you have the `NPM_TOKEN` environment variable defined."
+        );
+      }
+      npmConfig["//registry.npmjs.org/:_authToken"] = processEnv.NPM_TOKEN;
+      fs.writeFileSync(userNpmrcPath, ini.stringify(npmConfig));
```
We can assign new values to the config object and then let `npm/ini` serialize the content.

Andarist commented:
Thank you for the PR! This refactor might be worth it but it will require some further analysis - if you are fine with waiting then we can leave it as is and I will get back to it this week. I would prefer to merge a small PR, that would only fix the whitespace issue, first. That could hit the stable release sooner.

emmenko commented:
Sure, I opened a separate PR with only a quick fix for the regex: #116

Andarist commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,50 @@
+    }
+  } else {
+    console.log("No user .npmrc file found, creating one");
+    if (!processEnv.NPM_TOKEN) {
+      throw new Error(
+        "Missing NPM authToken. Please make sure you have the `NPM_TOKEN` environment variable defined."
+      );
+    }
```
This should be removed. There is also a possibility to have `.npmrc` file in the project itself - it takes precedence over the user-level `.npmrc` and this should be respected. The current logic is that we try to add token to the user-level `.npmrc` which should be harmless if there is a project-level `.npmrc` because it will "shadow" over the user-level `.npmrc` anyway.

emmenko replied:
Hmm ok but then we end up writing an "empty/undefined" `authToken` value. Is that what we want?

Andarist replied:
No, probably not - we can skip over creating/editing `.npmrc` when `NPM_TOKEN` is not set

Andarist commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,50 @@
+
+    let hasAuthToken = false;
+    for (const [key, value] of Object.entries(npmConfig)) {
+      if (/\/\/(.*)authToken$/.test(key) && Boolean(value)) {
```
We should tighten up this regex - one, in theory, can have defined multiple auth tokens, for multiple registries. If one has a token for registry foo then we should still create an entry for the npm registry (if it's missing)

emmenko replied:
Ah good point!

emmenko committed:
refactor: do not throw if NPM_TOKEN is missing

emmenko committed:
feat: add option to skipNpmrcCheck

emmenko commented on README.md:
```diff
@@ -10,6 +10,7 @@ This action for [Changesets](https://github.com/atlassian/changesets) creates a
 - version - The command to update version, edit CHANGELOG, read and delete changesets. Default to `changeset version` if not provided
 - commit - The commit message to use. Default to `Version Packages`
 - title - The pull request title. Default to `Version Packages`
+- skipNpmrcCheck - Check if there is a user-defined `.npmrc` file with a proper npm registry and an `authToken`. If not, a default `.npmrc` file is created.
```
@Andarist I added this option to opt-out of the .npmrc checks.

I find it useful in case I don't need the use the default npm registry (for example when using github packages).

What do you think?

emmenko commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,59 @@
+import * as ini from "ini";
+
+// https://docs.npmjs.com/using-private-packages-in-a-ci-cd-workflow#create-and-check-in-a-project-specific-npmrc-file
+const npmRegistryTokenKey = "//registry.npmjs.org/:_authToken";
```
I did some research and couldn't find a case where the value isn't exactly like this (for the npm registry). Therefore I think we can omit the regex and simply do a 1:1 check.

emmenko commented on src/npmUtils.ts:
```diff
@@ -0,0 +1,59 @@
+        fs.writeFileSync(userNpmrcPath, ini.stringify(npmConfig));
+      } else {
+        console.warn(
+          "Missing `NPM_TOKEN` environment variable, skipping update of .npmrc file."
```
So now in case the `NPM_TOKEN` is missing, we simply log a warning.

emmenko commented:
@Andarist Hey, I updated the PR based on the feedback. I also added a new option `skipNpmrcCheck` (see comment).

Let me know how you prefer to proceed here, thanks!

emmenko committed:
test: fix assertions
